### PR TITLE
Move buffer activity tracking to the core

### PR DESCRIPTION
**Pull request by justjanne**
## In Short
* Store the bitset of all message types that have occured in a buffer since the last seen message on the core
* Sync this with clients

## Status
- [x] Implement database backend
- [x] Integrate the activity tracking into the buffersyncer
- [x] Implement the actual sync of the activity between core and client
- [x] Assign a CoreFeatures flag
- [x] Have the client make use of this feature

## Rationale
Currently a client needs to request backlog for all channels, to know if events have occured on that channel (and potentially unhide it, show it differently in the UI, etc). This patch aims at removing that requirement by allowing the client to know the set of message types that have occured in a channel.

## Disadvantages
The core can not take ignore list items into account, so under some circumstances a channel might be marked as having new messages, but the client might not show them once the channel is activated. Additionally, this does not currently process highlights yet.

### CODE REVIEW

File: src/common/buffersyncer.cpp [hunk 1]
```diff
@@ -124,6 +125,35 @@ void BufferSyncer::initSetMarkerLines(const QVariantList &list)
 }
 
 
+QVariantList BufferSyncer::initActivities() const
+{
+    QVariantList list;
+    QHash<BufferId, Message::Types>::const_iterator iter = _bufferActivities.constBegin();
```
Comment (digitalcircuit) [inline src/common/buffersyncer.cpp]: Minor:  You can likely use...
```cpp
auto iter = _bufferActivities.constBegin();
```
...as per [`quassel.cpp`](https://github.com/quassel/quassel/blob/dd69349ca91776432a4a53aa4d18dd8ef018cd26/src/common/quassel.cpp#L498-L499 ).

File: src/core/sqlitestorage.cpp [hunk 2]
```diff
@@ -1475,6 +1475,79 @@ QHash<BufferId, MsgId> SqliteStorage::bufferMarkerLineMsgIds(UserId user)
...
+        QSqlQuery query(db);
+        query.prepare(queryString("select_buffer_bufferactivities"));
+        query.bindValue(":userid", user.toInt());
+
+        lockForRead();
+        safeExec(query);
+        error = !watchQuery(query);
+        if (!error) {
+            while (query.next()) {
+                bufferActivityHash[query.value(0).toInt()] = Message::Types(query.value(1).toInt());
```
Comment (digitalcircuit) [inline src/core/sqlitestorage.cpp]: I've poked through bits of Quassel's code yet can't seem to find this...

Does `Message::Types()` handle unknown values gracefully?
Is this a matter that would ever be a concern?  (E.g. adding/removing message types)

Comment (justjanne) [reply src/core/sqlitestorage.cpp]: Message::Types() is defined here: https://github.com/justjanne/quassel/blob/4b05ec5bd9c805ab316bd9053594d97df1d27131/src/common/message.h#L56 as QFlag<Message::Type>, which under the hood is just an int, with nicer operators for testing, | and &. It should handle it fine.

File: src/common/buffersyncer.cpp (hunk 1 above)
Comment (justjanne) [reply src/common/buffersyncer.cpp]: Should be addressed now

File: src/core/sqlitestorage.cpp (hunk 2 above)
Comment (digitalcircuit) [reply src/core/sqlitestorage.cpp]: Alright, that makes sense!  Thank you for explaining.

**Comment by digitalcircuit** [comment]
# Overview - 2017-9-22
## Summary
* SQLite and Postgres setup/migration **fails**
* Buffer activity updates work correctly **in some cases**, highlights still apply
* **Surface-level testing only**, will watch for issues after merge

## Steps for a successful test
1.  Initialize or upgrade Quassel core (*completing first run wizard if needed*)
2.  Verify networks intact
3.  Connect to a network, verify message sends **and buffer activity updates (including highlights)**
4.  Restart the core
5.  Verify networks intact **and buffer activity loads**

## Notes
* Surface level rather than extensive testing (adding/deleting, etc); I'll keep watch after this is in `master`
* As there's no change to setup flow, several tests reused the same initialized backend
* On first upgrade, existing channel activity is not detected
* Existing regression: `master` branch sometimes forgets highlighted state on client reconnect/core restart

# Test results - 🗷 fail
## Linux - 🗷 fail
### Fresh configuration - 🗷 fail
*Quassel core from this PR set up without any initial data*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, new core, new client | 🗹 *pass* |
SQLite, new core, old client | 🗹 *pass* | Reused initialized backend
SQLite, old core, new client | 🗷 **fail** | Does not detect channel activity, highlights work
SQLite, new monolithic | 🗹 *pass* |
Postgres, new core, new client | 🗹 *pass* |
Postgres, new core, old client | 🗹 *pass* | Reused initialized backend
Postgres, old core, new client | 🗹 *pass* |

### Migration - 🗷 fail
*Quassel core from this PR set up without any initial data, then migrated*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite → Postgres, new core, new client | 🗷 **fail** | *See Failure 1 below*

### Upgrading existing - 🗹 pass
*Quassel core 0.12.2 set up (0.13-git-master for monolithic), then upgraded to this PR*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, old → new core | 🗹 *pass* | Reused initialized backend
SQLite, old monolithic → new monolithic | 🗹 *pass* |
Postgres, old → new core | 🗹 *pass* | Reused initialized backend

## Windows - 🗷 fail
### Fresh configuration - 🗷 fail
*Quassel core from this PR set up without any initial data*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, new core, new client | 🗹 *pass* |
SQLite, new core, old client | 🗹 *pass* | Windows core with Linux client
SQLite, old core, new client | 🗷 **fail** | Does not detect channel activity, highlights work
SQLite, new monolithic | 🗹 *pass* |

### Migration - ? N/A
*Without Postgres on Windows, migration could not be tested*

### Upgrading existing - 🗹 pass
*Quassel core 0.12.4 set up, then upgraded to this PR*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, old → new core | 🗹 *pass* | Reused initialized backend
SQLite, old monolithic → new monolithic | 🗹 *pass* |

## Failed cases
### Failure 1, Linux, SQLite to Postgres
```
sly@vbox-quassel:~/Desktop$ sudo --user quasselcore quasselcore --select-backend=PostgreSQL --configdir=/var/lib/quassel
2017-09-22 23:05:23 Info: SQLite storage backend is ready. Schema version: 25 
2017-09-22 23:05:23 Info: Database authenticator is ready. 
Default values are in brackets
Username [quassel]: 
Password []: 
Hostname [localhost]: 
Port [5432]: 
Database [quassel]: 
unhandled Error in QSqlQuery!
                  last Query:
 SELECT value FROM coreinfo WHERE key = 'schemaversion'
              executed Query:
 SELECT value FROM coreinfo WHERE key = 'schemaversion'
                bound Values: 
                Error Number: 42601
               Error Message: ERROR:  syntax error at end of input
LINE 1: EXECUTE 
                ^
(42601) QPSQL: Unable to create query
              Driver Message: QPSQL: Unable to create query
                  DB Message: ERROR:  syntax error at end of input
LINE 1: EXECUTE 
                ^
(42601)
unhandled Error in QSqlQuery!
                  last Query:
 SELECT MAX(version) FROM coreinfo
              executed Query:
 SELECT MAX(version) FROM coreinfo
                bound Values: 
                Error Number: 42601
               Error Message: ERROR:  syntax error at end of input
LINE 1: EXECUTE 
                ^
(42601) QPSQL: Unable to create query
              Driver Message: QPSQL: Unable to create query
                  DB Message: ERROR:  syntax error at end of input
LINE 1: EXECUTE 
                ^
(42601)
Storage Schema is missing!
2017-09-22 23:05:28 Info: PostgreSQL storage backend is ready. Schema version: 24 
2017-09-22 23:05:28 Warning: Switched storage backend to: PostgreSQL 
Migrating storage backend SQLite to PostgreSQL...
Transferring QuasselUser...
Done.
Transferring Identity...
Done.
Transferring IdentityNick...
Done.
Transferring Network...
Done.
Transferring Buffer...
Migration Failed!
AbstractSqlMigrationReader::transferMo(): unable to transfer Migratable Object of type Buffer!
WriterError:
  executed Query:
INSERT INTO buffer (bufferid, userid, groupid, networkid, buffername, buffercname, buffertype, lastmsgid, lastseenmsgid, markerlinemsgid, bufferactivity, key, joined)
VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  bound Values:
0 :  1
1 :  1
2 :  0
3 :  1
4 :  #qcore_dev_test
5 :  #qcore_dev_test
6 :  2
7 :  66
8 :  66
9 :  66
10 :  0
11 :  
12 :  true
  Error Number: 42601
  Error Message: "ERROR:  syntax error at or near \"(\"\nLINE 1: EXECUTE  (1, 1, 0, 1, '#qcore_dev_test', '#qcore_dev_test', ...\n                 ^\n(42601) QPSQL: Unable to create query"
2017-09-22 23:05:28 Warning: Unable to migrate storage backend! (No migration writer for PostgreSQL)
```

**Comment by digitalcircuit** [comment]
# Overview - 2017-9-22, B
## Summary
* SQLite and Postgres setup/migration works
* Buffer activity updates work correctly, highlights still apply
* **Surface-level testing only**, will watch for issues after merge
* **Incomplete; skipped several tests** due to past testing/tiredness, should revisit later

## Steps for a successful test
1.  Initialize or upgrade Quassel core (*completing first run wizard if needed*)
2.  Verify networks intact
3.  Connect to a network, verify message sends **and buffer activity updates (including highlights)**
4.  Restart the core
5.  Verify networks intact **and buffer activity loads**

## Notes
* Surface level rather than extensive testing (adding/deleting, etc); I'll keep watch after this is in `master`
* As there's no change to setup flow, several tests reused the same initialized backend
* On first upgrade, existing channel activity is not detected
* Existing regression: `master` branch sometimes forgets highlighted state on client reconnect/core restart
* Skipped all non-Linux/non-Postgres tests other than migration due to past test; should revisit in future (*tired right now*)

# Test results - 🗹 pass (incomplete)
## Linux - 🗹 pass (incomplete)
### Fresh configuration - 🗹 pass (incomplete)
*Quassel core from this PR set up without any initial data*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, new core, new client | 🗹 *pass* |
SQLite, new core, old client | 🗹 *pass* | Reused initialized backend
SQLite, old core, new client | 🗹 *pass* |
SQLite, new monolithic | 🗹 *pass* |
Postgres, new core, new client | ? *skipped* |
Postgres, new core, old client | ? *skipped* | Reused initialized backend
Postgres, old core, new client | ? *skipped* |

### Migration - 🗹 pass
*Quassel core from this PR set up without any initial data, then migrated*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite → Postgres, new core, new client | 🗹 *pass* |

### Upgrading existing - 🗹 pass (incomplete)
*Quassel core 0.12.2 set up (0.13-git-master for monolithic), then upgraded to this PR*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, old → new core | ? *skipped* | Reused initialized backend
SQLite, old monolithic → new monolithic | ? *skipped* |
Postgres, old → new core | ? *skipped* | Reused initialized backend

## Windows - 🗹 pass (incomplete)
### Fresh configuration - 🗹 pass (incomplete)
*Quassel core from this PR set up without any initial data*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, new core, new client | ? *skipped* |
SQLite, new core, old client | ? *skipped* | Windows core with Linux client
SQLite, old core, new client | ? *skipped* | Does not detect channel activity, highlights work
SQLite, new monolithic | ? *skipped* |

### Migration - ? N/A
*Without Postgres on Windows, migration could not be tested*

### Upgrading existing - 🗹 pass (incomplete)
*Quassel core 0.12.4 set up, then upgraded to this PR*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, old → new core | ? *skipped* | Reused initialized backend
SQLite, old monolithic → new monolithic | ? *skipped* |

## Failed cases
None found!

File: src/core/SQL/PostgreSQL/select_buffer_bufferactivities.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+SELECT bufferid, bufferactivity
+FROM buffer
+WHERE userid = :userid
```
Comment (Sput42) [inline src/core/SQL/PostgreSQL/select_buffer_bufferactivities.sql]: Newline at EOF missing

File: src/core/SQL/PostgreSQL/select_buffer_bufferactivity.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+SELECT COALESCE(SUM(t.type),0)
+FROM
+  (SELECT DISTINCT TYPE
+   FROM backlog
+   WHERE bufferid = :bufferid
+     AND messageid > :lastseenmsgid) t;
```
Comment (Sput42) [inline src/core/SQL/PostgreSQL/select_buffer_bufferactivity.sql]: Newline at EOF missing

File: src/core/SQL/PostgreSQL/update_buffer_bufferactivity.sql [hunk 5]
```diff
@@ -0,0 +1,3 @@
+UPDATE buffer
+SET bufferactivity = :bufferactivity
+WHERE userid = :userid AND bufferid = :bufferid
```
Comment (Sput42) [inline src/core/SQL/PostgreSQL/update_buffer_bufferactivity.sql]: Newline at EOF

File: src/core/SQL/PostgreSQL/version/24/upgrade_000_alter_buffer_add_bufferactivity.sql [hunk 6]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE buffer
+ADD COLUMN bufferactivity integer NOT NULL DEFAULT 0
```
Comment (Sput42) [inline src/core/SQL/PostgreSQL/version/24/upgrade_000_alter_buffer_add_bufferactivity.sql]: Newline at EOF

File: src/core/SQL/SQLite/select_buffer_bufferactivities.sql [hunk 7]
```diff
@@ -0,0 +1,3 @@
+SELECT bufferid, bufferactivity
+FROM buffer
+WHERE userid = :userid
```
Comment (Sput42) [inline src/core/SQL/SQLite/select_buffer_bufferactivities.sql]: Newline at EOF

File: src/core/SQL/SQLite/select_buffer_bufferactivity.sql [hunk 8]
```diff
@@ -0,0 +1,6 @@
+SELECT COALESCE(SUM(t.type),0)
+FROM
+  (SELECT DISTINCT TYPE
+   FROM backlog
+   WHERE bufferid = :bufferid
+     AND messageid > :lastseenmsgid) t;
```
Comment (Sput42) [inline src/core/SQL/SQLite/select_buffer_bufferactivity.sql]: Newline at EOF

File: src/core/SQL/SQLite/update_buffer_bufferactivity.sql [hunk 9]
```diff
@@ -0,0 +1,3 @@
+UPDATE buffer
+SET bufferactivity = :bufferactivity
+WHERE userid = :userid AND bufferid = :bufferid
```
Comment (Sput42) [inline src/core/SQL/SQLite/update_buffer_bufferactivity.sql]: Newline at EOF

File: src/core/SQL/SQLite/version/25/upgrade_000_alter_buffer_add_bufferactivity.sql [hunk 10]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE buffer
+ADD COLUMN bufferactivity integer NOT NULL DEFAULT 0
```
Comment (Sput42) [inline src/core/SQL/SQLite/version/25/upgrade_000_alter_buffer_add_bufferactivity.sql]: Newline at EOF

File: src/core/core.h [hunk 11]
```diff
@@ -531,6 +531,39 @@ class Core : public QObject
...
 
+    //! Update the BufferActivity for a Buffer
+    /** This Method is used to make the activity state of a Buffer persistent
+     *  \note This method is threadsafe.
+     *
+     * \param user      The Owner of that Buffer
+     * \param bufferId  The buffer id
+     * \param MsgId     The Message id where the marker line should be placed
+     */
+    static inline void setBufferActivity(UserId user, const BufferId &bufferId, const Message::Types &activity) {
```
Comment (Sput42) [inline src/core/core.h]: Ints and enums shouldn't be passed as const refs

File: src/core/core.h [hunk 12]
```diff
@@ -531,6 +531,39 @@ class Core : public QObject
...
+    }
+
+    //! Get the bitset of buffer activity states for a buffer
+    /** This method is used to load the activity state of a buffer when its last seen message changes.
+     *  \note This method is threadsafe.
+     *
+     * \param bufferId The buffer
+     * \param lastSeenMsgId     The last seen message
+     */
+    static inline Message::Types bufferActivity(BufferId &bufferId, const MsgId &lastSeenMsgId) {
```
Comment (Sput42) [inline src/core/core.h]: Int arguments shouldn't be refs (and surely not a non-const ref, since this isn't an outparam...)

File: src/core/corebuffersyncer.cpp [hunk 13]
```diff
@@ -34,17 +34,20 @@ class PurgeEvent : public QEvent
...
 {
+    connect(parent, SIGNAL(displayMsg(Message)), SLOT(addBufferActivity(Message)));
 }
 
 
 void CoreBufferSyncer::requestSetLastSeenMsg(BufferId buffer, const MsgId &msgId)
 {
-    if (setLastSeenMsg(buffer, msgId))
+    if (setLastSeenMsg(buffer, msgId)) {
+        setBufferActivity(buffer, (int) Core::bufferActivity(buffer, msgId));
```
Comment (Sput42) [inline src/core/corebuffersyncer.cpp]: In new code I'd prefer avoiding C-style casts (use `static_cast<int>()` instead)

File: src/core/corebuffersyncer.cpp [hunk 14]
```diff
@@ -181,3 +189,8 @@ void CoreBufferSyncer::purgeBufferIds()
         }
     }
 }
+
+void CoreBufferSyncer::setBufferActivity(BufferId buffer, const int &activity) {
```
Comment (Sput42) [inline src/core/corebuffersyncer.cpp]: This shouldn't  be a const ref, and possibly it should be using the actual type (`Message::Types`) instead?

File: src/core/corebuffersyncer.h [hunk 15]
```diff
@@ -48,7 +57,10 @@ public slots:
 
     virtual void requestPurgeBufferIds();
 
-    virtual inline void requestMarkBufferAsRead(BufferId buffer) { markBufferAsRead(buffer); }
+    virtual inline void requestMarkBufferAsRead(BufferId buffer) {
+        setBufferActivity(buffer, (int) Message::Types());
```
Comment (Sput42) [inline src/core/corebuffersyncer.h]: You should be able to just write `{}` to use the default-constructed value (instead of `(int)Message::Types()`)

File: src/core/postgresqlstorage.cpp [hunk 16]
```diff
@@ -1385,6 +1385,61 @@ QHash<BufferId, MsgId> PostgreSqlStorage::bufferMarkerLineMsgIds(UserId user)
     return markerLineHash;
 }
 
+
+void PostgreSqlStorage::setBufferActivity(UserId user, const BufferId &bufferId, const Message::Types &bufferActivity)
```
Comment (Sput42) [inline src/core/postgresqlstorage.cpp]: Integral types shouldn't be passed as const ref

File: src/core/postgresqlstorage.h [hunk 17]
```diff
@@ -96,6 +96,9 @@ public slots:
     virtual QHash<BufferId, MsgId> bufferLastSeenMsgIds(UserId user);
     virtual void setBufferMarkerLineMsg(UserId user, const BufferId &bufferId, const MsgId &msgId);
     virtual QHash<BufferId, MsgId> bufferMarkerLineMsgIds(UserId user);
+    virtual void setBufferActivity(UserId id, const BufferId &bufferId, const Message::Types &type);
+    virtual QHash<BufferId, Message::Types> bufferActivities(UserId id);
+    virtual Message::Types bufferActivity(BufferId &bufferId, const MsgId &lastSeenMsgId);
```
Comment (Sput42) [inline src/core/postgresqlstorage.h]: Ah, here I see where all the const refs come from. So one could argue that you're being consistent, however I'd prefer that newly written code should be written properly (of course, one could also fix the existing code to be consistent :))

File: src/common/buffersyncer.h [hunk 18]
```diff
@@ -45,9 +53,19 @@ public slots:
     QVariantList initMarkerLines() const;
     void initSetMarkerLines(const QVariantList &);
 
+    QVariantList initActivities() const;
+    void initSetActivities(const QVariantList &);
+
     virtual inline void requestSetLastSeenMsg(BufferId buffer, const MsgId &msgId) { REQUEST(ARG(buffer), ARG(msgId)) }
     virtual inline void requestSetMarkerLine(BufferId buffer, const MsgId &msgId) { REQUEST(ARG(buffer), ARG(msgId)) setMarkerLine(buffer, msgId); }
 
+    virtual inline void setBufferActivity(BufferId buffer, const int &activity) {
```
[the remark this answers is not in the document]
Comment (justjanne) [reply src/common/buffersyncer.h]: Message::Types can not be deserialized by older clients/cores. If they receive it, they close the connection.

We could simply avoid sending this message to clients that don't understand it, but that is a bit more complicated (and requires some more, newer filtering code we don't have available yet)

[the remark this answers is not in the document]
Comment (justjanne) [reply src/common/buffersyncer.h]: Specifically, if we use int, we get on older clients/cores

> "no matching slot for sync call: BufferSyncer::setBufferActivity (objectName=\"\"). Params are:" (QVariant(BufferId, ), QVariant(int, 0))

and if we use a type the client/core doesn't know, we might end up with it closing the connection instead

**Comment by justjanne** [comment]
I believe there is still a bug in this code with new queries and activities, I'll have to take a look at that.

File: src/common/buffersyncer.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (Sput42) [reply src/common/buffersyncer.h]: Ah ok, that sounds reasonable.

File: src/uisupport/clickable.cpp [hunk 19]
```diff
@@ -61,7 +61,7 @@ ClickableList ClickableList::fromString(const QString &str)
     static QRegExp regExp[] = {
         // URL
         // QRegExp(QString("((?:https?://|s?ftp://|irc://|mailto:|www\\.)%1+|%1+\\.[a-z]{2,4}(?:?=/%1+|\\b))%2").arg(urlChars, urlEnd)),
-        QRegExp(QString("\\b(%1%2(?:/%3*)?)%4").arg(scheme, authority, urlChars, urlEnd), Qt::CaseInsensitive),
+        QRegExp(QString("\\b((?:(?:mailto:|(?:[+.-]?\\w)+://)|www(?=\\.\\S+\\.))(?:(?:[,.;@:]?[-\\w]+)+\\.?|\\[[0-9a-f:.]+\\])(?::\\d+)?(?:/(?:[,.;:]*[\\w~@/?&=+$()!%#*-])*)?)(?:>|[,.;:\"]*\\s|\\b|$)").arg(scheme, authority, urlChars, urlEnd), Qt::CaseInsensitive),
```
Comment (Sput42) [inline src/uisupport/clickable.cpp]: This looks fishy , and also like it should not be part of this commit at all...
